### PR TITLE
 [serve] Fix deploy config edge case bug

### LINKED ISSUE #40324: [Serve] Bug when redeploying old config
### What happened + What you expected to happen

If you do the following:
1. Deploy an app that succeeds
2. Redeploy the app with different runtime env / import path / args which fails upon importing / building the app
3. Redeploy the app using the same config from step 1 (should succeed)

Step 3 will fail with an error:
```
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/serve/_private/application_state.py", line 340, in deploy_config
    overrided_infos = override_deployment_info(
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/serve/_private/application_state.py", line 967, in override_deployment_info
    info = deployment_infos[deployment_name]
TypeError: 'NoneType' object is not subscriptable
```

### Versions / Dependencies

2.7

### Reproduction script

Follow steps from above.

### Issue Severity

None

### PR DESCRIPTION
**Pull request by zcin**
## Why are these changes needed?

See issue

## Related issue number

Closes https://github.com/ray-project/ray/issues/40324

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by edoakes — APPROVED** [review]

**Comment by edoakes** [comment]
@zcin some (seemingly unrelated) test failures, merging master to re-trigger CI

**Review by sihanwang41 — APPROVED** [review]

File: python/ray/serve/_private/application_state.py [hunk 1]
```diff
@@ -368,7 +368,11 @@ def deploy_config(
                 ray.cancel(self._build_app_task_info.obj_ref)
 
             # Halt reconciliation of target deployments
-            self._set_target_state_deployment_infos(None)
+            self._set_target_state(
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/application_state.py:371]: Is it possible for this edge case to occur when the other calls to `self._set_target_state_deployment_infos(None)` happen? Specifically, [here](https://github.com/ray-project/ray/blob/c525af0b7e9cc89fffdb8ecea9f98c8a7b0c45bb/python/ray/serve/_private/application_state.py#L348) and [here](https://github.com/ray-project/ray/blob/c525af0b7e9cc89fffdb8ecea9f98c8a7b0c45bb/python/ray/serve/_private/application_state.py#L353)?

**Review by shrekris-anyscale — APPROVED** [review]

File: python/ray/serve/_private/application_state.py (hunk 1 above)
Comment (zcin) [reply python/ray/serve/_private/application_state.py:371]: Good catch, yeah that same edge case can happen if the config override options failed. Let me push an update to fix this too, thanks

**Comment by zcin** [comment]
@edoakes Tests are passing, ready for merge!

**Pull request merged**